circuit/algorithms/src/keccak/hash.rs

```diff
@@ -0,0 +1,454 @@
+    fn permutation_f<const WIDTH: usize, const NUM_ROUNDS: usize>(
+        input: Vec<Boolean<E>>,
+        round_constants: &[U64<E>],
+        rotl: &[usize],
+    ) -> Vec<Boolean<E>> {
+        debug_assert_eq!(input.len(), WIDTH, "The input vector must have {WIDTH} bits");
+        debug_assert_eq!(
+            round_constants.len(),
+            NUM_ROUNDS,
+            "The round constants vector must have {NUM_ROUNDS} elements"
+        );
+
+        // Partition the input into 64-bit chunks.
+        let mut a = input.chunks(64).map(U64::from_bits_le).collect::<Vec<_>>();
+        // Permute the input.
+        for round_constant in round_constants.iter().take(NUM_ROUNDS) {
+            a = Self::round(a, round_constant, rotl);
+        }
+        // Return the permuted input.
+        a.into_iter().flat_map(|e| e.to_bits_le()).collect()
+    }
+
+    /// The round function `R` is defined as follows:
+    /// ```text
+    /// R = ι ◦ χ ◦ π ◦ ρ ◦ θ
+    /// ```
+    /// where `◦` denotes function composition.
+    fn round(a: Vec<U64<E>>, round_constant: &U64<E>, rotl: &[usize]) -> Vec<U64<E>> {
+        debug_assert_eq!(a.len(), MODULO * MODULO, "The input vector 'a' must have {} elements", MODULO * MODULO);
+
+        const WEIGHT: usize = MODULO - 1;
+
+        /* The first part of Algorithm 3, θ:
+         *
+         * for x = 0 to 4 do
+         *   C[x] = a[x, 0]
+         *   for y = 1 to 4 do
+         *     C[x] = C[x] ⊕ a[x, y]
+         *   end for
+         * end for
+         */
+        let mut c = Vec::with_capacity(WEIGHT);
+        for x in 0..MODULO {
+            c.push(&a[x] ^ &a[x + MODULO] ^ &a[x + (2 * MODULO)] ^ &a[x + (3 * MODULO)] ^ &a[x + (4 * MODULO)]);
+        }
+
+        /* The second part of Algorithm 3, θ:
+         *
+         * for x = 0 to 4 do
+         *   D[x] = C[x−1] ⊕ ROT(C[x+1],1)
+         *   for y = 0 to 4 do
+         *     A[x, y] = a[x, y] ⊕ D[x]
+         *   end for
+         * end for
+         */
+        let mut d = Vec::with_capacity(WEIGHT);
+        for x in 0..MODULO {
+            d.push(&c[(x + 4) % MODULO] ^ Self::rotate_left(&c[(x + 1) % MODULO], 63));
+        }
+        let mut a_1 = Vec::with_capacity(WEIGHT * WEIGHT);
+        for y in 0..MODULO {
+            for x in 0..MODULO {
+                a_1.push(&a[x + (y * MODULO)] ^ &d[x]);
+            }
+        }
+
+        /* Algorithm 4, π:
+         *
+         * for x = 0 to 4 do
+         *   for y = 0 to 4 do
+         *     (X, Y) = (y, (2*x + 3*y) mod 5)
+         *     A[X, Y] = a[x, y]
+         *   end for
+         * end for
+         *
+         * Algorithm 5, ρ:
+         *
+         * A[0, 0] = a[0, 0]
+         * (x, y) = (1, 0)
+         * for t = 0 to 23 do
+         *   A[x, y] = ROT(a[x, y], (t + 1)(t + 2)/2)
+         *   (x, y) = (y, (2*x + 3*y) mod 5)
+         * end for
+         */
+        let mut a_2 = a_1.clone();
```

Comment:
```suggestion
        let mut a_2 = a;
```
we can reuse the `a` vector here to avoid this clone

Reply:
I don't think this is true